README.md
# Kaggle Titanic challenge
Here is my implementation of the Kaggle Titanic challenge. My purpose was to use this challenge
to do my first Python script, my first Machine Learning work, and a practical use of Pandas and
Scikit-Learn.

Most of the models currently implemented allowed to me to hit 78.9% accuracy on test data provided
by Kaggle, so if you have something else using this, well there is a problem somewhere! I mostly 
used the Support Vector Classifier for this task, despite most of peoples on Kaggle reporting better 
results using a Decision Tree. Well for me it was not the case. You can test yourself, there is an 
option to train a Decision Tree.

I will detail the content in subsequent sections, but would like to make a summary of the files here:
- all the text files are monitoring outputs from the code. You can skip them, there is no code inside.
- the main part of the code (the executable script) is **titanic.py**. This is where all options and
models are defined.
- the directory **titalib** contains custom-made librairies. Those weren't *necessary* but greatly
help improving the readability of the script and its versatility. The feature engineering is performed
into one of the scripts from this directory, we will come back later on it.
- the **data** directory contains data downloaded from the Kaggle challenge page, as well as the 
prediction file produced by the code.

I will not discuss feature engineering here as I wrote a Jupyter notebook for this purpose 
(*analysis.ipynb*).

## Requirements
I created this script using:
- Python 3.6.2
- Scikit-Learn 0.19.0
- Pandas 0.20.3

## Usage
Simply execute the script to train the default SVC with the parameter C=2 and use a Radial Basis Function
as a kernel of parameter gamma=0.0173. It will produce a prediction file formatted to be submitted on 
Kaggle under the `data/` directory. You can go into `titanic.py` script to change those settings.

All the features are transformed into categorical features, one-hot-encoded (for non-binary ones) and
are scaled to mean=0 and std=1. Currently, there are 5 categorical features left (*Age*, *Fare*, *Sex*, 
*Relatives* and *Location*), 3 being binary (*Age*, *Sex* and *Relatives*), one with 3 categories (*Fare*)
and the last with 4 categories (*Location*). This give a total of 96 combinations, which is about one order
of magnitude below the data count (which is about 800). We then can reasonably hope that each of our 
combination will be populated, and that our models will be able to train on each possible case. The former is not
necessarly true, but I still think it is preferable to train on as many configurations as possible.

Alternatively, you might want to play around with other models, or explore different parameters settings.
There are numerous flags for this. The full option list is detailed below. They are all defined into 
`titanic.py` script. You can also go for a different feature engineering route. For this, you will have 
to go into the `dataformat()` function in the `data/preproc.py` file.

Paths:
- `--train_data` will specify the full path to the training file.
- `--test_data` will specify the full path to the test file.
- `--pred_output` will specify the full path to the prediction file.

Mode (accepts 0 or 1):
- `--grid`: wether or not you want to run a grid search on the hyperparameters. The dictionaries
still have to be written inside `titanic.py` by yourself if you want to change the default grid search settings
I used.
- `--test`: I just wanted to be able to perform quick testings on models an manually split the data into
a cross-validation set and a training set after shuffling the data. Its purpose is more about testing code 
rather than models so you can totally skip this section.
- `--multi`: the multi-mode is a simple ensemble method in this code. It is used to produce a prediciton file
made from the equal-weight-averaged predictions from a k-Nearest Neighbors, a Support Vector Classifier, and a 
Logistic Regression. Scikit-learn provides functions to do that, but I didn`t know at that time.
- `--model`: this option specify the model you want to use for your grid search, your testing or if you run
in single mode (the default option, that's to say `grid`, `test` and `multi` are all set to 0).
The available models are for instance a Support Vector Classifier ( `SVC` ), a Decision Tree ( `Tree` ),
 a Logistic Regression ( `LR` ) or k-Nearest Neighbors ( `KNN` ). By default, this option is set to `SVC`. Also 
 note that this field is append to the output text files.

Note that except for the model option, those modes are all exclusive to each other.

Data settings:
- `--child`: the limit in years below which a passenger will be treated as a child.
- `--sibsp`: the maximum number of vanilla categories preserved for *SibSp* feature after label-encoding. 
All numbers above in the data will be put into the same additional category. I don`t use this feature anymore 
so changing this option will have no effect unless you remove it from the list of features to be drop.
- `--parch`: same as above but for the *Parch* features. Both those features have been combined into a new one 
called *Relatives*. As above, it needs to be removed from the list of features to be drop if you want this option 
to be of any use.
- `--fare1`: the upper limit of the "poor" category of passengers.
- `--fare2`: the lower limit of the "rich" category of passengers. A third category lies between those two.
- `--title`: the minimum number of occurences of a given title to be placed into its own category. Anything below 
that will be put into the same category.

## The files
### titanic.py
This is the main script. You will find inside all the options to run the code (described above) and the settings
of data slicing. The most important things you need to know here are about the following lines, near the beginning 
of the code:

- `droplist = ['Name','Ticket','Survived','Cabin','Embarked','Pclass','SibSp','Parch']`  
The droplist is the list of pandas columns to be kicked after the data have been pre-processed. You see here
that the features *SibSp* and *Parch* are on the droplist, that's why their cut option defined above is currently
useless. If you want them back, remove them from this list and add instead *Relatives*, which is a new feature built
on those two.

- `mask = [False,True,False,False,True]`  
The mask indicates whether or not a column has to be one-hot encoded or not. Binary features shouldn't foo obvious
reasons, as well as numerical features. To know precisely which fields corresponds to which feature, simply print
the first element of the pandas dataframe output by the `prep.dataformat()` function.

### titalib/preproc.py
This files contains functions related to the pre-processing of data, namely the feature engineering, the one-hot
encoding and the feature scaling. The `dataformat()` function is in charge to perform the feature engineering
and is probably the most important of all the files if you had to choose one to read.

### titalib/models.py
This files contains functions to run the different operating modes described above. Those are just shortcuts
to train a model, print infos to terminal and write an output file in order to make the main script lighter.

### titalib/printer.py
This files contains the routines to write all the output files, and basic infos about the chosen settings 
in the terminal.

### textfiles/grid_output_*.txt
Those files give more detailed results of the grid search provided by scikit-learn for a given model. It will 
give you all the parameters combination tested, their rank (the first being the best parameters combination), 
as well as the mean train/test score and the associated standard deviation for each data split used.

### textfiles/cv_log_*.txt
This files simply contains the full output of scikit-learn's grid search ( the `cv_results_` attribute of
the `GridSearchCV` object), in case you need more informations.

### textfiles/fails_*.txt
The funnier of output files. This one contains a list of all the data entries the given model fails to 
reproduce. Can be usefull before deciding or not to use an ensemble method and which weight to use for their
vote, or simply to understand why the current model fails (if it is because of outliers, if there is a particular
pattern etc...).


history.txt
# C     gamma     limits                    test  train   kaggle  NOTES
# 4.22  0.0083    7,2,1,10,100,30           0.851 0.862
# 5.66  0.0072    7,2,1,10,100,30           0.850 0.863   0.73
# 5.33  0.0077    15,2,1,10,50,10           0.836 0.841   0.770
# 4.84  0.0081    15,2,1,10,50,10,10        0.828 0.850   0.775
# 5.05  0.0079    15,2,1,10,50,5,10         0.825 0.853   0.770
# 7.16  0.0079    4,2,1,10,50,10,10         0.839 0.891   0.737   age as slices
# 4.21  0.0084    18,2,1,10,50,10,10        0.828 0.849   0.775   
# 6.89  0.0061    15,2,1,10,50,10,50        0.834 0.843   0.794   
# 39.7  0.0027    15,2,1,10,50,10,50        0.835 0.846   0.780   age nan as means(title) + title feature
# 2.89  0.0095    15,2,1,10,50,10,50        0.834 0.839   0.775   ticket as fare norm, age as slices
# 6.89  0.0061    15,2,1,10,50,10,50        0.851 -----   0.746   add KNN output as new feature
  5-uniform       
# 3.84  0.0086    15,2,1,10,50,10,50        0.850 -----   0.761  
  3-uniform       
# 3.84  0.0086    15,2,1,10,50,10,50        0.872 -----   0.741 
  3-distance        
# 3.84  0.0086    15,2,1,10,50,10,50        0.877 -----   0.756 
  6-distance        
# 20.0  0.0015    15,2,1,10,50,10,50        0.879 0.876   0.756   found best param by mixing grid search
  6-distance        
# 0.12  'l2'      15,2,1,10,50,10,50        0.814 0.824   0.760   logistic regression
# 15.51  0.0019   15,2,1,10,50,10,50        0.879 0.876   0.737   averaged pred over KNN, SVC and LR
  5-uniform       
  0.12  'l2'        
# 6.00  0.0120    15,2,1,10,50,10,50        0.838 0.845   0.756   add SVC pred as feature of LR
  1.00  'l2'        
# 6.80  0.0060    15,2,1,10,50,10,50        0.832 -----   0.761   
  0.20  'l2'        
# 3.78  0.0010    15,3,2,10,40,50,10        0.821  0.823  0.779   vote system (KNN,SVC,LR)
  18-uniform                                0.824  0.837
  0.07  'l2'                                0.802  0.814
# 3.00  0.0009    15,3,2,10,40,50,10        0.821  0.823  0.785
  18-uniform                                0.824  0.837
  0.12  'l2'                                0.822  0.830
# 3.00  0.0009    15,3,2,10,40,50,10        0.821  0.823  0.780
# 18-uniform      15,3,2,10,40,50,10        0.824  0.837  0.780
# 0.12  'l2'      15,3,2,10,40,50,10        0.822  0.830  0.780
    
# 8.73  0.0038    16,40,2,2,10,50,20,100    0.824  0.831  0.780   Age NaN filled (Name detailed processing)
# 3.00  0.0070    16,40,2,2,3,10,50,20,100  0.824  0.828  0.789   SibSp and Parch replaced by Relatives
# 7.33  0.0050    16,40,2,2,3,10,50,20,100  0.825  0.829  0.780   
# 6.40  0.0050    15,50,2,2,3,10,50,20,100  0.827  0.829  0.789  
# 8.63  0.0114    15,50,2,2,3,10,50,20,100  0.818  0.838  0.765   class_weight='balanced'
# 8.44  0.0050    15,50,2,2,3,10,50,20,100  0.828  0.828  -----   reduced Embarkment to 2 categories
# 2.00  0.0137    16,2,2,3,15,35,20         0.827  0.827  0.789   remove Cabin, Embarkment, last Age bound 
# 2.00  0.0137    16,2,2,4,15,35,20         0.831  0.834  0.789   
# 2.00  0.0137    16,2,2,4,15,35,20         0.831  0.834  0.789   vote system (KNN,SVC,LR)
  0.27  'l1'      
  16-distance    
# 5.50  0.0064    16,2,2,4,15,50,20         0.829  0.831  0.789   remove Pclass, creates Location
# 4.00  0.0064    16,2,2,15,50,20           0.822  0.823  0.789   Hard code Relatives to 2 categories
# 4.00  0.0064    16,2,2,15,50,20           0.822  0.823  0.789   Reduce Location to 4 categories
# 1.00  0.0173    15,2,2,15,50,20           0.826  0.826  0.789   
# 2.00  0.0173    15,2,2,15,50,20           0.825  0.826  0.789   


titanic.py
"""
================================
Titanic challenge draft
================================

v.091017
"""
print(__doc__)

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import argparse

import titalib.preproc as prep
import titalib.printer as prt
import titalib.models as mdl

from sklearn import svm
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.linear_model import LogisticRegression
from sklearn import neighbors
from sklearn import tree

########################
# PARSE INPUTS
########################
parser = argparse.ArgumentParser(description='Define important inputs.')

# Path section
parser.add_argument("--train_data",
                    type=str,
                    default='data/train.csv',
                    help="Path to the training data")
parser.add_argument("--test_data",
                    type=str,
                    default='data/test.csv',
                    help="Path to the test data")
parser.add_argument("--pred_output",
                    type=str,
                    default='data/predictions.csv',
                    help="Path to output predictions file")

# Options section
parser.add_argument("--grid",
                    type=int,
                    default=0,
                    help="Switch to activate(=1)/deactivate(=0) grid search")
parser.add_argument("--test",
                    type=int,
                    default=0,
                    help="activate(=1)/deactivate(=0) test mode for manual data splitting")
parser.add_argument("--multi",
                    type=int,
                    default=0,
                    help="activate(=1)/deactivate(=0) multi model mode")
parser.add_argument("--model",
                    type=str,
                    default='SVC',
                    help="model to be used for single-model mode ('SVC', 'LR', 'KNN', 'Tree')")

# Settings section
parser.add_argument("--child",
                    type=int,
                    default=15,
                    help="Limit of age for childrens.")
parser.add_argument("--sibsp",
                    type=int,
                    default=2,
                    help="Number of categories for SibSp to be kept (see 'preproc.dataformat')")
parser.add_argument("--parch",
                    type=int,
                    default=3,
                    help="Number of categories for Parch to be kept (see 'preproc.dataformat')")
parser.add_argument("--fare1",
                    type=float,
                    default=25,
                    help="Upper limit of lowest fare category.")
parser.add_argument("--fare2",
                    type=float,
                    default=50,
                    help="Lower limit of highest fare category.")
parser.add_argument("--title",
                    type=int,
                    default=20,
                    help="Min of data required for Title field (see 'preproc.dataformat')")
                    
FLAGS, unparsed = parser.parse_known_args()
limits = {"child": FLAGS.child,
          "sibsplimit": FLAGS.sibsp,
          "parchlimit": FLAGS.parch,
          "farebound1": FLAGS.fare1,
          "farebound2": FLAGS.fare2,
          "titlelimit": FLAGS.title}


########################
# DATA PRE-PROCESSING
########################
# load the data
data_train_raw = prep.dataload(FLAGS.train_data)
data_test_raw = prep.dataload(FLAGS.test_data)

# save index of test set as a list for reconstruction of ouput file
index = data_test_raw.index.tolist()

# fields to be kicked after data procesing
droplist = ['Name','Ticket','Survived','Cabin','Embarked','Pclass','SibSp','Parch']
# format and clean up the data
fullset, trainsize, labels = prep.dataformat(data_train_raw,
                                             data_test_raw,
                                             droplist=droplist,
                                             limits=limits)

# mask of categorical (non-binary) features to be one-hot-encoded
mask = [False,True,False,False,True]
# one-hot encoding and split back data 
data_train, data_test, labels = prep.dataonehot(dataset=fullset,
                                                mask=mask,
                                                labels=labels,
                                                trainsize=trainsize)

if FLAGS.model != 'Tree':
  # mean normalization to mean=0 std=1
  data_train = prep.datanorm(data_train)
  data_test = prep.datanorm(data_test)

# model and parameters definitions
classif = svm.SVC(cache_size=2000)
params_SVC = {'C': 2, 'gamma': 0.0173, 'kernel': 'rbf'}

knear = neighbors.KNeighborsClassifier(algorithm='kd_tree')
params_KNN = {'n_neighbors': 5, 'weights': 'distance' }

logistic = LogisticRegression()
params_LR = {'C': 0.27, 'penalty': 'l1' }

dtree = tree.DecisionTreeClassifier()
params_Tree = {}
  
###############################
# grid search on the parameters
###############################
if FLAGS.grid == 1:
  # prepare the 'shuffle and split' tool
  n_splits = 5
  n_repeats = 10
  test_size = 1/n_splits
  rkf = RepeatedStratifiedKFold(n_splits=n_splits, n_repeats=n_repeats, random_state=None)
  
  # parameters to be tested
  param_C = np.linspace(1,10,10)
  param_Gam = np.linspace(0.001,0.05,10)
  params_SVC = [{'C': param_C, 'gamma': param_Gam, 'kernel': ['rbf']}]
  
  param_neighbors = range(2,20)
  params_KNN = [{'n_neighbors': param_neighbors, 'weights': ['uniform', 'distance']}]
  
  param_C2 = np.linspace(0.0001,1,20)
  params_LR = [{'C': param_C2, 'penalty': ['l2', 'l1']}]
  
  params_Tree = {}

  # grid search
  if FLAGS.model == 'SVC':
    dummy = mdl.gridsearch(model = classif,
                           switch = FLAGS.model,
                           params = params_SVC,
                           X_train = data_train,
                           y_train = labels,
                           splits = rkf,
                           test_size = test_size,
                           n_repeats = n_repeats)
  elif FLAGS.model == 'KNN':
    dummy = mdl.gridsearch(model = knear,
                           switch = FLAGS.model,
                           params = params_KNN,
                           X_train = data_train,
                           y_train = labels,
                           splits = rkf,
                           test_size = test_size,
                           n_repeats = n_repeats)
  elif FLAGS.model == 'LR':
    dummy = mdl.gridsearch(model = logistic,
                           switch = FLAGS.model,
                           params = params_LR,
                           X_train = data_train,
                           y_train = labels,
                           splits = rkf,
                           test_size = test_size,
                           n_repeats = n_repeats)
  elif FLAGS.model == 'Tree':
    dummy = mdl.gridsearch(model = dtree,
                           switch = FLAGS.model,
                           params = params_Tree,
                           X_train = data_train,
                           y_train = labels,
                           splits = rkf,
                           test_size = test_size,
                           n_repeats = n_repeats)
  else:
    print("No model selected. Please choose one with the --model option.")
            
########################################################
# Fitting and predictions (no-grid, failed pred outputs)
########################################################
elif FLAGS.test == 1:
  # manually split into test/train sets
  n_samples = len(data_train_raw)
  testsize = int(0.3*n_samples)

  np.random.seed(int(np.random.uniform(0,100,1)[0]))
  order = np.random.permutation(n_samples)
  data_train = data_train[order]
  labels = labels[order]

  X_train = data_train[testsize:]
  y_train = labels[testsize:]
  X_test = data_train[:testsize]
  y_test = labels[:testsize]
                  
  # train each model separately
  if FLAGS.model == 'SVC':
    mdl.traintest(model = classif,
                  switch = FLAGS.model,
                  params = params_SVC,
                  X_train = X_train,
                  y_train = y_train,
                  X_test = X_test,
                  y_test = y_test,
                  testsize = testsize,
                  order = order)
  elif FLAGS.model == 'KNN':                    
    mdl.traintest(model = knear,
                  switch = FLAGS.model,
                  params = params_KNN,
                  X_train = X_train,
                  y_train = y_train,
                  X_test = X_test,
                  y_test = y_test,
                  testsize = testsize,
                  order = order)
  elif FLAGS.model == 'LR':                  
    mdl.traintest(model = logistic,
                  switch = FLAGS.model,
                  params = params_LR,
                  X_train = X_train,
                  y_train = y_train,
                  X_test = X_test,
                  y_test = y_test,
                  testsize = testsize,
                  order = order)
  elif FLAGS.model == 'Tree':                  
    mdl.traintest(model = dtree,
                  switch = FLAGS.model,
                  params = params_Tree,
                  X_train = X_train,
                  y_train = y_train,
                  X_test = X_test,
                  y_test = y_test,
                  testsize = testsize,
                  order = order)
  else:
    print("No model selected. Please choose one with the --model option.")
      
##############################################
# Fitting and predictions outputs (multi mode)
##############################################
elif FLAGS.multi == 1:
  # train a KNN, SVC, LR on the data
  knn_pred = mdl.trainpred(model = knear,
                           switch = FLAGS.model,
                           params = params_KNN,
                           X_train = data_train,
                           y_train = labels,
                           X_test = data_test)
  svm_pred = mdl.trainpred(model = classif,
                           switch = FLAGS.model,
                           params = params_SVC,
                           X_train = data_train,
                           y_train = labels,
                           X_test = data_test)
  lr_pred = mdl.trainpred(model = logistic,
                          switch = FLAGS.model,
                          params = params_LR,
                          X_train = data_train,
                          y_train = labels,
                          X_test = data_test)
                           
  # average the predictions of the different models
  predictions = ((knn_pred + svm_pred + lr_pred)/3 + 0.5).astype(int)
  
  # score on training test
  pred = ((knear.predict(data_train) + classif.predict(data_train) + logistic.predict(data_train))/3 + 0.5).astype(int)
  score = 1 - ((abs(labels-pred)).sum() / len(data_train))
  print("Total training score: %0.3f" % score)
  
  # Write the prediction file
  prt.outwriter(filename = FLAGS.pred_output,
                testindex = index,
                predictions = predictions)


###############################################
# Fitting and predictions outputs (single mode)
###############################################
else:
  # train a single model on the data
  if FLAGS.model == 'SVC':
    predictions = mdl.trainpred(model = classif,
                                switch = FLAGS.model,
                                params = params_SVC,
                                X_train = data_train,
                                y_train = labels,
                                X_test = data_test)
  elif FLAGS.model == 'KNN':                             
    predictions = mdl.trainpred(model = knear,
                                switch = FLAGS.model,
                                params = params_KNN,
                                X_train = data_train,
                                y_train = labels,
                                X_test = data_test)
  elif FLAGS.model == 'LR':                               
    predictions = mdl.trainpred(model = logistic,
                                switch = FLAGS.model,
                                params = params_LR,
                                X_train = data_train,
                                y_train = labels,
                                X_test = data_test)
  elif FLAGS.model == 'Tree':                               
    predictions = mdl.trainpred(model = dtree,
                                switch = FLAGS.model,
                                params = params_Tree,
                                X_train = data_train,
                                y_train = labels,
                                X_test = data_test)
  else:
    print("No model selected. Please choose one with the --model option.")  
    
  # Write the prediction file
  prt.outwriter(filename = FLAGS.pred_output,
                testindex = index,
                predictions = predictions)
  

